pkg/applicationserver/applicationserver.go

```diff
@@ -588,6 +592,202 @@ func (as *ApplicationServer) handleJoinAccept(ctx context.Context, ids ttnpb.End
+		logger = logger.WithField("pending_session_key_id", pendingSession.SessionKeyID)
+	}
+	currentSession := dev.Session
+	if currentSession != nil {
+		logger = logger.WithField("current_session_key_id", currentSession.SessionKeyID)
+	}
+	ctx = log.NewContext(ctx, logger)
+	oldPendingLastAFCntDown := pendingSession.GetLastAFCntDown()
+	oldCurrentLastAFCntDown := currentSession.GetLastAFCntDown()
+	defer func() {
+		// If something fails, clear the downlink queue as an empty downlink queue is better than a downlink queue
+		// with items that are encrypted with the wrong AppSKey.
+		if err != nil {
+			if pendingSession != nil {
+				pendingSession.LastAFCntDown = oldPendingLastAFCntDown
+			}
+			if currentSession != nil {
+				currentSession.LastAFCntDown = oldCurrentLastAFCntDown
+			}
+			logger.WithError(err).Warn("Recalculate downlink queue failed; clearing the downlink queue")
+			client := ttnpb.NewAsNsClient(link.conn)
+			req := &ttnpb.DownlinkQueueRequest{
+				EndDeviceIdentifiers: dev.EndDeviceIdentifiers,
+			}
+			if _, err := client.DownlinkQueueReplace(ctx, req, link.callOpts...); err != nil {
+				logger.WithError(err).Warn("Failed to clear the downlink queue; any queued items in the Network Server are invalid")
+				events.Publish(evtInvalidQueueDataDown(ctx, dev.EndDeviceIdentifiers, err))
+			} else {
+				events.Publish(evtLostQueueDataDown(ctx, dev.EndDeviceIdentifiers, err))
+			}
+		}
+	}()
+	if dev.SkipPayloadCrypto {
+		return errPayloadCryptoDisabled
+	}
+	newCurrentQueue, newPendingQueue, err := f(ctx, dev)
+	if err != nil {
+		return err
+	}
+	downlinks := append(newCurrentQueue[:0:0], newCurrentQueue...)
+	downlinks = append(downlinks, newPendingQueue...)
+	client := ttnpb.NewAsNsClient(link.conn)
+	req := &ttnpb.DownlinkQueueRequest{
+		EndDeviceIdentifiers: dev.EndDeviceIdentifiers,
+		Downlinks:            downlinks,
+	}
+	_, err = client.DownlinkQueueReplace(ctx, req, link.callOpts...)
+	return err
+}
+
+var errUnknownSession = errors.DefineNotFound("unknown_session", "unknown session")
+
+// recomputePendingQueue partitions the invalidated downlinks based on their session key ID. Downlinks which are part of the
+// current session are not modified, while the downlinks from the previous pending session are migrated to the new pending
+// session.
+// This method requires the given end device's pending session to be set. This method mutates the end device's pending session LastAFCntDown.
+// This method does not change the contents of the given invalid downlink queue.
+func (as *ApplicationServer) recomputePendingQueue(ctx context.Context, dev *ttnpb.EndDevice, link *link, previousPendingSession *ttnpb.Session, invalidatedDownlinks []*ttnpb.ApplicationDownlink) error {
+	return as.changeDownlinkQueue(ctx, dev, link, func(ctx context.Context, dev *ttnpb.EndDevice) ([]*ttnpb.ApplicationDownlink, []*ttnpb.ApplicationDownlink, error) {
+		var previousPendingQueue, currentQueue []*ttnpb.ApplicationDownlink
```

Comment:
This should simply do `migrated, err := as.migrateDownlinkQueue(ctx, dev.EndDeviceIdentifiers, invalidatedDownlinks, dev.Session, dev.PendingSession, 1)` and `return invalidatedQueue, migrated, nil`

There's no need for this method, in fact, just put this at `L582`